### ISSUE TITLE
Wrong declaration order in generated output

### ISSUE THREAD

**Issue by jacereda** [carp-lang/Carp#526]
The following code:

```
(load "Vector.carp")

(deftype (V f) [v (Vector3 f)])
(deftype (X f) [x (Maybe (V f))])

(defn main []
  (let [x (X.init (the (Maybe (V Float)) (Maybe.Nothing)))]
    0))
```

yields:

```
./out/main.c:445:5: error: unknown type name 'Maybe__V__float'
    Maybe__V__float x;
    ^
1 error generated.
```

because the generated file declares types in the wrong order:

```
...
// Depth 103
typedef struct {
    Maybe__V__float x;
} X__float;
...
// Depth 105
typedef struct {
    union {
    struct {
        V__float member0;
    } Just;
    // Nothing
    };
    char _tag;
} Maybe__V__float;
...
```

**Comment by jacereda** [comment]
Could it be a problem in `scoreTypeBinder`?

**Comment by eriksvedang** [comment]
Probably, yes!

**Comment by jacereda** [comment]
I think I have a fix, is https://github.com/carp-lang/Carp/pull/529 going to be merged? If so I can PR it on top of it.

**Comment by jacereda** [comment]
OTOH, it's a trial and error fix. Shouldn't the scoring be something like a `Data.Graph.topSort` of the dependencies of each binding?

**Comment by eriksvedang** [comment]
Sorry, I'm moving houses (and having a vacation) so I haven't had the time to merge PR:s, sorry.

**Comment by eriksvedang** [comment]
I merged #529 now, but don't have time to look at the others until in a couple of days.

**Comment by jacereda** [comment]
No hurries, I just wanted to know on top of which branch to PR this one. Enjoy your vacation.
